### PR TITLE
GitHub action

### PR DESCRIPTION
**Pull request by Eomm**
Working on #42 

Need to:

- support NPM_TOKEN instead of `--npm-otp` and verify that it is not conflicting

### CODE REVIEW

**Comment by Eomm** [comment]
Hi @salmanm 
I saw you contributed to a lot to the fastify org automation! Thanks for that, this is great!

I would like to ask you if are you interested to help me to push forward this PR?

I'm not so good to create GH actions so I would really appreciate your hints on this scenario:
right now we release (almost) every module with this CLI (releasify) run locally.
The key features are: npm publishing and GH Releases with changelog.

We would like to run it on GH, so it would be even easier to release a version.

Right now I'm not so sure that I made some good choices since:
- this PR would add an ACTION together with a CLI and maybe a dedicated repo would be better/clearer
- we adopt 2FA auth on npm to release that has blocked this PR at the beginning so I saw some modules that ask for it when releasing (via ngrok) _I will search that module if needed_
- whatever you think may help to improve this process - even cut away this CLI

Feel free to ignore this message if I overwhelmed you and sorry if I have annoyed you somehow 😄

File: .github/action/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+FROM node:12-alpine
+# The action uses the last released version of releasify
+RUN npm install releasify -g && releasify -v
```
Comment (zekth) [inline .github/action/Dockerfile:3]: why installing releasify as we already have it as we are in the repo? why not just `COPY . ./` ? I mean also you can target the version if we use proper tagging.

**Comment by salmanm** [comment]
Hi @Eomm, sure, would be happy to assist. Here's my 2c.

But quick question, how did you plan to bypass the `whoamI` check on the CI? I mean `releasify` cli runs `npm whoami` and fails if there's no npm login. May be I missed something, may be there's another way, please let me know.

Assuming there is a way to work around that, I don't think we need a dedicated action or even fire up docker to run this command. We can simply create a new workflow in the repo. Let me know if you've already discarded this idea for some reason.

```yml
name: release

on:
  workflow_dispatch:
    inputs:
      # ... all inputs (semver, etc ...)

jobs:
  release:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
        with:
          node-version: 12
      - run: npm install
      - run: npm i releasify
      - run: |
          ./node_modules/.bin/releasify \
            publish \
            --verbose debug \
            --semver ${{github.event.inputs.semver}} \
            --remote ${{github.event.inputs.remote}} \
            --branch ${{github.event.inputs.branch}} \
            --major ${{github.event.inputs.major}} \
            --gh-token ${{secrets.GITHUB_TOKEN}}
```

**Comment by zekth** [comment]
You need to use a token to be passed to the action. See : https://docs.npmjs.com/using-private-packages-in-a-ci-cd-workflow

eg: https://github.com/JS-DevTools/npm-publish

**Comment by Eomm** [comment]
> Let me know if you've already discarded this idea for some reason.

No! I think it was just overengineering and I didn't mind to run your simple example and it looks much better and easy to use!
Regarding the `whoami` I thought that the npm token would have solved the issue.
Right now I didn't play with tokens really so this PR was less than a draft.

Anyway, I like your approach and I can focus on trying it, many thanks!

> You need to use a token to be passed to the action

Do you know if the token will ask for an OTP number? I expect so
I used tokens in past, but without 2FA as we do

**Comment by zekth** [comment]
> > You need to use a token to be passed to the action
> 
> Do you know if the token will ask for an OTP number? I expect so
> I used tokens in past, but without 2FA as we do

I need to investigate on this

**Comment by salmanm** [comment]
> Regarding the whoami I thought that the npm token would have solved the issue.

It does for `npm` but releasify seems to have an explicit check for `whoami` which fails. 

> Do you know if the token will ask for an OTP number?

You can make do without OTP if you use npm automation token.

Screenshot_20201121-061553__01

However, I heard @mcollina saying that it's not safe. _Which I'd like to learn how._

**Comment by Eomm** [comment]
I will work on this today so I will write down my findings 👍🏼
Thank you for the hints! 

Regarding the token without otp, it is unsafe because a malicious dependency could leak the process.env vars and steal the tokens.

So, for example, we should create one token for every repo to minimise the impact and check deeply all the dependencies tree.
I think the main reason is this one

**Comment by jsumners** [comment]
I really think that hack for automated publishing tokens is a dumb idea by NPM. It, in my opinion, effectively removes the 2FA setting altogether.

**Comment by Eomm** [comment]
> I really think that hack for automated publishing tokens is a dumb idea by NPM. It, in my opinion, effectively removes the 2FA setting altogether.

Yes, indeed and I'm working to keep the 2FA active.

I have tried for now (without success) to run this publish:
https://github.com/Eomm/node-workshop-micro/runs/1435866378?check_suite_focus=true

As recap:

- `whoami` must be turned off to work with CI
- I have a publish error due to:
  - the token permission (even if I have double-checked it ) 
   - or based on the scoped module I was using - I will try without the scope
- it will be necessary to create a flow to input the OTP (funniest thing)

I feel I'm closer to the solution thanks to the input received here 🔝

**Comment by salmanm** [comment]
@Eomm Looks like the error is not due to token permission or scope. It's because `npm` need auth creds present in `.npmrc`.

If you add the following steps before "release now" step, I think it should work.

```yml
     - run: npm config set //registry.npmjs.org/:_authToken=${NPM_TOKEN}
        env:
          NPM_TOKEN: ${{secrets.NPM_TOKEN}}
    - run: npm whoami
```

With this approach, we don't even have to turn off `whoami` check in releasify. Does that help?

**Comment by salmanm** [comment]
> Regarding the token without otp, it is unsafe because a malicious dependency could leak the process.env vars and steal the tokens.

@Eomm @jsumners Just for my knowledge (and please excuse me if the question is very basic).
When exactly can a dependency steal envs? while doing `npm i` step?
Nevertheless, as per my understanding, the envs can be read by steps **only** when we supply them explicitly using `env:` key (like above). They can't access the secrets directly.

**Comment by jsumners** [comment]
> When exactly can a dependency steal envs? while doing `npm i` step?

I'm not sure what this is referring to. My point is that if the token used for publish ever leaks, e.g. via a mistaken GitHub Action log, then the integrity of the published module is no longer secure. Which is as effective as create a bot account with 2FA disabled.

**Comment by Eomm** [comment]
Thank you @salmanm ! I thought the env parameter was enough!

It works now: we have a basic flow action! it is not so beautiful but it works! 😄 
I will work to move the OTP input from the action to an async step since if the flow is too much slow, the publish could fail.

> When exactly can a dependency steal envs? 

Yes, if we don't remember to put `--no-script` arg
I was not aware that if the `env:` keyword is not in the `run:` the step doesn't receive it! So running

```
- run npm i
- run echo 'foo'
```

the `run npm i` will have an empty `process.env` right?

**Comment by salmanm** [comment]
> the run npm i will have an empty process.env right?

Yep, exactly. 🙂

**Comment by salmanm** [comment]
@jsumners 

> My point is that if the token used for publish ever leaks, e.g. via a mistaken GitHub Action log

I see, so it's more of a precaution than "actual" hazard. Fair enough. I just didn't think of this as an issue because GitHub automatically redacts secrets printed to the log. 
Thanks

**Comment by Eomm** [comment]
> because GitHub automatically redacts secrets printed to the log.

Confirmed, in the linked actions run, you may spot the `***` in the raw logs

> I see, so it's more of a precaution than "actual" hazard

I think it so right now, but it will be funny to add this async input 😈

**Comment by jsumners** [comment]
The list of caveats around secrets redaction in Actions is a long one -- https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/security-hardening-for-github-actions

**Comment by Eomm** [comment]
In the [GH docs](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idsteps) I read

> Each step runs in its own process in the runner environment and has access to the workspace and filesystem

So, since the file system could be read by modules (instead of ENV var) it could be a potential issue

**Comment by Eomm** [comment]
Closing this PR since we don't need a GH Action, but only a GH Workflow

I'm working on it and I would continue the discussion in the issue #42

**Pull request closed without merging**